java/fury-core/src/main/java/org/apache/fury/Fury.java

```diff
@@ -1221,15 +1240,55 @@ public Object deserializeJavaObjectAndClass(FuryReadableChannel channel) {
     return deserializeJavaObjectAndClass(buf);
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public <T> T copy(T obj) {
     if (Objects.isNull(obj)) {
       return null;
     }
-    if (obj instanceof FuryCopyable) {
-      return (T) ((FuryCopyable) obj).copy(this);
-    } else {
-      return (T) classResolver.getOrUpdateClassInfo(obj.getClass()).getSerializer().copy(obj);
+    try {
+      copyDepth++;
+      if (Objects.isNull(originToCopyMap)) {
+        originToCopyMap = new IdentityMap();
+      }
+      Object newObj = originToCopyMap.get(obj);
```

Comment:
For objects like String, Integer, there is no need to track ref, the map put/get cost is expensive for small objects copy

Reply:
You can take Serializer#needToTrackRef as an example